include/unifex/detach_on_cancel.hpp

```diff
@@ -0,0 +1,212 @@
+
+  struct cancel_callback final {
+    detached_state* state_;
+
+    void operator()() noexcept {
+      state_->stopSource_.request_stop();
+      auto* op = state_->parentOp_.exchange(nullptr, std::memory_order_acq_rel);
+      if (op != nullptr) {
+        // We won the race for cancellation
+        // ownership is transferred to the detached state
+        (void)op->state_.release();
+        op->callback_.destruct();
+        unifex::set_done(std::move(op->receiver_));
+      }
+    }
+  };
+
+  struct type final {
+    type(UpstreamSender&& s, DownstreamReceiver&& r)
+      : receiver_(std::move(r))
+      , state_(std::make_unique<detached_state>(this, (UpstreamSender &&) s)){};
+
+    friend void tag_invoke(
+        tag_t<unifex::start>,
+        typename operation_state<UpstreamSender, DownstreamReceiver>::type&
+            op) noexcept {
+      auto& childOp = op.state_->childOp_;
+      op.callback_.construct(
+          get_stop_token(op.receiver_), cancel_callback{op.state_.get()});
+      unifex::start(childOp);
+    }
+
+    remove_cvref_t<DownstreamReceiver> receiver_;
+    manual_lifetime<typename stop_token_type_t<
+        DownstreamReceiver>::template callback_type<cancel_callback>>
+        callback_;
+    std::unique_ptr<detached_state> state_;
+  };
+};
+
+template <typename UpstreamSender, typename DownstreamReceiver>
+struct operation_state<UpstreamSender, DownstreamReceiver>::_receiver final {
+  auto tryGetOp() noexcept {
+    // if cancellation won, parentOp_ would be null at this point
+    // see operation_state::cancel_callback()
+    auto* op = state_->parentOp_.exchange(nullptr, std::memory_order_acq_rel);
+    if (op != nullptr) {
+      op->callback_.destruct();
+    } else {
+      delete state_;
+    }
+    return op;
+  }
+
+  template <typename... Values>
+  void set_value(Values&&... values) noexcept {
+    if (auto op = tryGetOp()) {
+      UNIFEX_TRY {
+        unifex::set_value(std::move(op->receiver_), (Values &&) values...);
+      }
+      UNIFEX_CATCH(...) {
+        unifex::set_error(std::move(op->receiver_), std::current_exception());
+      }
+    }
+  }
+
+  template <typename Error>
+  void set_error(Error&& error) noexcept {
+    if (auto op = tryGetOp()) {
+      unifex::set_error(std::move(op->receiver_), (Error &&) error);
+    }
+  }
+
+  void set_done() noexcept {
+    if (auto op = tryGetOp()) {
+      unifex::set_done(std::move(op->receiver_));
+    }
+  }
+
+  friend inplace_stop_token
+  tag_invoke(tag_t<get_stop_token>, const _receiver& r) noexcept {
+    return r.state_->stopSource_.get_token();
+  }
+  detached_state* state_;
+};
+
+template <typename UpstreamSender, typename DownstreamReceiver>
+struct operation_state<UpstreamSender, DownstreamReceiver>::detached_state
+    final {
+  detached_state(
+      typename operation_state<UpstreamSender, DownstreamReceiver>::type* op,
+      UpstreamSender&&
+          s) noexcept(is_nothrow_connectable_v<UpstreamSender, _receiver>)
```

Comment:
Some weird formatting that breaks a single-character parameter name onto a separate line here.
clang-format's just being weird?